### DISCUSSION TITLE
2024/04/shiny-for-python/

### DISCUSSION THREAD

**Discussion by giscus** [rbind/xiangyun#102] [General]
# 2024/04/shiny-for-python/

https://xiangyun.rbind.io/2024/04/shiny-for-python/

**Comment by chuxinyuan** [comment]
“萼”字楷体在手机上显示不正常，按照[益辉的说法](https://github.com/yihui/yihui.org/discussions/1547)：

> 唯一可能的问题是你的文章里用了我从没[用过](https://github.com/yihui/cron/blob/main/fonts/chars.txt)的生僻字，但这个可能性应该很小。如果真有这种情况，你按[我的脚本](https://github.com/yihui/cron/blob/main/.github/workflows/fonts.yaml)自己去编就好了。

**Reply by XiangyunHuang** [reply]
看着挺复杂的，我就不多折腾了。

**Comment by jeremy-feng** [comment]
谢谢分享！我想请教一下，我使用

```python
fig.write_image("img/iris.png")
```

将带有中文标题的图片导出到本地表格时，打开图片后，发现其中的中文是方块。

![1732086676957](https://github.com/user-attachments/assets/e56d805b-301d-4cb2-a8a5-2e2da8d4ba0c)

请问博主是怎么实现 plotly 导出带有中文的图片的呢，是否需要进行什么配置？

**Reply by XiangyunHuang** [reply]
奇怪，我也不知道。当时我用的是macOS系统，配置好后没有什么问题。你的系统是啥？或者你看看 Plotly 的 GitHub 仓库的 issues，也看看是不是软件版本的问题。

**Reply by jeremy-feng** [reply]
谢谢！我的系统是 Linux，我再看看有没有解决方案 :smile:

**Reply by XiangyunHuang** [reply]
那你有安装中文字体吗？

**Reply by jeremy-feng** [reply]
应该是有的，因为如果我不导出图片而只是在 Jupyter Notebook 中展示的话，是能看到中文的，甚至还能更换中文字体。但是一旦导出图片后就变成方块了。

**Reply by jeremy-feng** [reply]
我在两台不同的 Linux 机器上进行了测试。两个机器都能在 Jupyter Notebook 中展示中文字体，但只有一个能导出带有中文字体的静态图片，另一个会将中文显示为方块。我检查发现，前者机器有 `/usr/share/fonts` 文件夹，而后者没有。

同时，我也找到了几个相关的 issue：
https://github.com/plotly/Kaleido/issues/157
https://community.plotly.com/t/using-kaleido-cannot-write-chinese-character-properly-but-ok-in-jupyter-lab/56586

这几个 issue 都与用于导出静态图片的 Kaleido 有关，因此我推测应该是 Kaleido 找不到字体文件的问题。

另外，plotly 本身是可以正确处理中文字体的，因此在 Jupyter Notebook 中可以正确显示中文，并且 `write_html` 也可以正确显示中文。但用 Kaleido 导出静态文件时就不能显示中文了。
